lib/facilities/bulk_client.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+require 'facilities/gis_middleware'
+
+# Client class for doing periodic bulk refreshes of GIS data.
+# Allows inspection of last modified date to update only when needed.
+module Facilities
+  class BulkClient
+    BATCH_SIZE = 1000.0
+    OPEN_TIMEOUT = 2
+    REQUEST_TIMEOUT = 6
+
+    COUNT_PARAMS = {
+      where: '1=1',
+      returnCountOnly: true,
+      f: 'json'
+    }.freeze
+
+    ALL_PARAMS = {
+      where: '1=1',
+      inSR: 4326,
+      outSR: 4326,
+      returnGeometry: true,
+      outFields: '*',
+      f: 'json'
+    }.freeze
+
+    def initialize(url)
+      @url = url
+      @conn = Faraday.new(url: @url) do |conn|
+        conn.options.open_timeout = OPEN_TIMEOUT
+        conn.options.timeout = REQUEST_TIMEOUT
+        conn.response :gis_json
+        # conn.response :logger
+        # TODO conn.use :breaker
+        conn.adapter Faraday.default_adapter
+      end
+    end
+
+    def last_edit_date
+      response = @conn.get('', f: 'json')
+      response.body.dig('editingInfo', 'lastEditDate')
+    rescue Faraday::ConnectionFailed => e
+      Rails.logger.warn "Facility lastEditDate check connection failed: #{e.message}"
+      return nil
+    rescue Faraday::TimeoutError
+      Rails.logger.warn 'Facility lastEditDate check timeout'
+      return nil
+    rescue Facilities::Errors::ServiceError => e
+      Rails.logger.warn "Facility lastEditDate check request failed: #{e.message}"
+      return nil
+    end
+
+    def fetch_all
+      Rails.logger.info "Refreshing from #{@url}"
+      query_url = [@url, 'query'].join('/')
+      count = get_count(query_url)
+      fetch_in_batches(query_url, count) unless count.nil?
+    rescue Faraday::ConnectionFailed => e
+      Rails.logger.error "Facility fetch connection failed: #{e.message}"
+      raise Facilities::Errors::ServiceError, 'Facility connection failed'
+    rescue Faraday::TimeoutError
+      Rails.logger.error 'Facility fetch timeout'
+      raise Facilities::Errors::ServiceError, 'Facility connection timed out'
+    rescue Facilities::Errors::ServiceError => e
+      Rails.logger.error "Facility fetch request failed: #{e.message}"
+      raise
+    end
```

Comment:
Ideally these rescue blocks would be handled in some sort of a "perform" block, all in one place rather than duplicated many times over. I realize that the messages are distinct, but I wonder if that's necessary if you have a stacktrace in Sentry for debugging purposes.